### ISSUE TITLE
NestJS with webpack HMR - TsMorph error

### ISSUE THREAD

**Issue by rmaes4** [mikro-orm/mikro-orm#523]
**Describe the bug**
I can't seem to get MikroORM working with NestJS. I have been referencing the MikroORM / NestJS real world app example. The only thing I can think of that I am doing differently vs the example is that I am using HMR.

**Stack trace**
```
[Nest] 41163   - 04/28/2020, 9:33:40 PM   [ExceptionHandler] Cannot read property 'getInstanceProperties' of undefined +89ms
TypeError: Cannot read property 'getInstanceProperties' of undefined
    at TsMorphMetadataProvider.readTypeFromSource (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/TsMorphMetadataProvider.js:55:59)
    at async TsMorphMetadataProvider.initPropertyType (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/TsMorphMetadataProvider.js:45:36)
    at async TsMorphMetadataProvider.initProperties (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/TsMorphMetadataProvider.js:30:17)
    at async TsMorphMetadataProvider.loadEntityMetadata (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/TsMorphMetadataProvider.js:19:9)
    at async MetadataDiscovery.discoverEntity (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/MetadataDiscovery.js:131:13)
    at async Function.runSerial (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/utils/Utils.js:391:22)
    at async MetadataDiscovery.findEntities (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/MetadataDiscovery.js:60:13)
    at async MetadataDiscovery.discover (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/metadata/MetadataDiscovery.js:31:9)
    at async Function.init (/Users/robmaes/Documents/ccp/api/node_modules/mikro-orm/dist/MikroORM.js:45:24)
    at async Injector.instantiateClass (/Users/robmaes/Documents/ccp/api/node_modules/@nestjs/core/injector/injector.js:293:37)
    at async callback (/Users/robmaes/Documents/ccp/api/node_modules/@nestjs/core/injector/injector.js:76:30)
    at async Injector.resolveConstructorParams (/Users/robmaes/Documents/ccp/api/node_modules/@nestjs/core/injector/injector.js:117:24)
    at async Injector.loadInstance (/Users/robmaes/Documents/ccp/api/node_modules/@nestjs/core/injector/injector.js:80:9)
    at async Injector.loadProvider (/Users/robmaes/Documents/ccp/api/node_modules/@nestjs/core/injector/injector.js:37:9)
    at async Promise.all (index 4)
    at async InstanceLoader.createInstancesOfProviders (/Users/robmaes/Documents/ccp/api/node_modules/@nestjs/core/injector/instance-loader.js:42:9)
```

**To Reproduce**
1. I have the following folder structure

- src/
    - cat/
    - cat.controller.ts
    - cat.entity.ts
    - cat.module.ts
    - cat.service.ts
- app.module.ts
- main.ts

2. app.module.ts
```ts
MikroOrmModule.forRoot({
      type: 'postgresql',
      host: process.env.DATABASE_HOST,
      port: parseInt(process.env.DATABASE_PORT),
      user: process.env.DATABASE_USERNAME,
      password: process.env.DATABASE_PASSWORD,
      dbName: process.env.DATABASE_DATABASE,
      entities: [Cat],
      debug: true,
      forceUtcTimezone: true
    })
```
3. cat.module.ts
```ts
@Module({
  imports: [MikroOrmModule.forFeature({ entities: [Cat] })],
  providers: [CatService],
  controllers: [CatController],
  exports: [CatService]
})
export class CatModule {}
```
4. cat.service.ts
```ts
@Injectable()
export class CatService {
  constructor(
    @InjectRepository(Cat) private readonly catRepository: EntityRepository<Cat>
  ) {}
}
```
5. cat.entity.ts
```ts
import { Entity, Enum, PrimaryKey, Property } from 'mikro-orm';
import { v4 } from 'uuid';

enum Species {
  LION = 'lion',
  TIGER = 'tiger',
  PANTHER = 'panther',
  CHEETAH = 'cheetah',
  LEOPARD = 'leopard'
}

@Entity()
export class Cat {
  @PrimaryKey()
  uuid = v4();

  @Property()
  createdAt = new Date();

  @Property({ onUpdate: () => new Date() })
  updatedAt = new Date();

  @Property()
  name!: string;

  @Property()
  age?: number;

  @Enum()
  species!: Species;
}

```

6. Project is setup according to [this NestJS guide](https://docs.nestjs.com/recipes/hot-reload)

7. Run `nest build --webpack --webpackPath webpack-hmr.config.js` to start the local server

**Expected behavior**
MikroORM will work.

**Additional context**
Following the Nest conventions I would like to keep entity files in their respective modules. e.g. cat.entity.ts in the cat folder. Therefore I don't have a single 'entities' folder I can point MikroORM to. I'd also prefer to not have to manually add the path for each one of my entities. I did not have to do that when I was using TypeORM.

I am really curious about this ORM, hence why I am trying to test it out coming from TypeORM. If someone could point me in the right direction that would be much appreciated. Thanks in advance!

**Versions**

| Dependency | Version |
| - | - |
| node | 12.14.1 |
| typescript | ^3.7.4 |
| mikro-orm | ^3.6.12 |
| your-driver | ^8.0.2 |

**Comment by Yaty** [comment]
https://mikro-orm.io/docs/deployment/#deploy-a-bundle-of-entities-and-dependencies-with-webpack should help you

**Comment by B4nan** [comment]
TsMorphMetadataProvider (the default one) is not compatible with webpack. Will add a warning to this place. 

There was also a suggestion to introduce different way to handle this - via custom TS transformer (https://github.com/cevek/ttypescript), will need to investigate this later, currently focusing on v4 development.

**Comment by rmaes4** [comment]
Ok thank you all for your help. I added
```ts
{
    ...
    discovery: {
        disableDynamicFileAccess: true
    }
}
```
to my config options and the server will now startup.

Is there an equivalent setting to `synchronize: true` like I was using before with Nest? This setting automatically creates the tables and columns etc to match the entity classes. If not I have a question about the schema cli:

So, before, I was using the Nest ConfigModule.

I had a file config/database.config.ts
```ts
export default registerAs(
  'database',
  (): Options => ({
    type: 'postgresql',
    host: process.env.DATABASE_HOST,
    port: parseInt(process.env.DATABASE_PORT),
    user: process.env.DATABASE_USERNAME,
    password: process.env.DATABASE_PASSWORD,
    //database: process.env.DATABASE_DATABASE,
    dbName: process.env.DATABASE_DATABASE,
    //synchronize: true,
    //autoLoadEntities: true,
    //keepConnectionAlive: true
    entities: [Cat],
    entitiesDirsTs: ['src'],
    debug: true,
    forceUtcTimezone: true
  })
);
```

and then in my app.module.ts
```ts
@Module({
  imports: [
    ConfigModule.forRoot({
      load: [databaseConfig, authConfig, redisConfig, microserviceConfig]
    }),
    ...
```
```ts
...
MikroOrmModule.forRootAsync({
      imports: [ConfigModule],
      useFactory: (configService: ConfigService) =>
        configService.get('database'),
      inject: [ConfigService]
    }),
```

This all works fine. But assuming I have to use the schema CLI  I'm unsure as to how I can get it to read my config file and load my .env file. I'd rather not hard code the values. I'd also rather not create two config files or have to use dotenv and the Nest config dotenv simultaneously. 

Any advice is appreciated. Thanks for the speedy responses.

**Comment by B4nan** [comment]
You can control the schema generator programatically, so if you want such behaviour, you could call `await orm.getSchemaGenerator().updateSchema();` right after you initialize the ORM. There is also `ensureDatabase()` method. 

There is also safe mode, which disabled dropping of tables and table column, which would be probably good to use here, so you don't loose any data. 

I would strongly advice not to use this in production. By doing this you also loose the ability to use migrations, which should be the preferred way to handle schema changes.

https://mikro-orm.io/docs/schema-generator#using-schemagenerator-programmatically

**Comment by rmaes4** [comment]
Thank you again for your reply. I will take a look at the link you sent. I don't intend to use this in production. I'm in the early development stage and I find it saves time and helps me iterate faster.

One other question (sorry!). I hard coded my DB credentials and ran the CLI schema creator. I noticed that it created my UUID and Date fields as "json" columns in my postgres DB. Did I do something wrong or do these types need to be specified manually? Is there a list of all the available "columnType: " options? Thanks!

**Comment by B4nan** [comment]
That will be because of the `ReflectMetadataProvider` you are now using, there are some limitations:

https://mikro-orm.io/docs/metadata-providers/#limitations-and-requirements

Namely you will need to provide the type explicitly when you have default values, so `uuid: string = v4()` and `createdAt: Date = new Date();`.

**Comment by rmaes4** [comment]
I decided to create a new Nest project for testing. This time I went ahead and made a dedicated entities folder. I also did not configure webpack HMR. Following your advice I created an app.service.ts file

```ts
@Injectable()
export class AppService implements OnModuleInit {
  constructor(private readonly orm: MikroORM) {}
  async onModuleInit() {
    const generator = this.orm.getSchemaGenerator();
    await generator.updateSchema();
  }
}
```

Everything seems to be working. My entities now automatically sync with the DB. I think I'm ready to now start migrating from TypeORM. Thank you for all of your help.
